### PR TITLE
[wip] Adds Docker configuration for an image proxy (closes #104).

### LINKED ISSUE #104: Use an image proxy to resize large images and prevent IP address leakage.
Not clear if we can reuse the embedly-proxy for this purpose or not, let's look into it.

### PR DESCRIPTION
**Pull request by chuckharmston**
This PR contains working configuration for a small image proxy that will do a few important things for us:

- Proxy images to outbound services to prevent leaking users' IP addresses to those service providers
- Resize (+ cache) unpredictably large images coming to us from service providers
- Allow us to use H/2 multiplexing (or, less ideally, H/1.1 keep-alive) to speed up those images. This would require our reverse proxy to serve this container at the same hostname as our application.

@mostlygeek, @relud: this isn't necessary for public launch next week, but we would like to get this in place somewhat quickly afterward. How can we make it easiest for you to manage and deploy this?

## TODO

- [ ] Document the purpose and usage; blocked on major changes to local dev documentation in #129 
- [ ] Move `conf/imageproxy/build.sh` to the new script in #129
- [ ] Allow whitelist to be configured with an environment variable

### CODE REVIEW

**Comment by 6a68** [comment]
@chuckharmston Is this a temporary addition until we solve the shared image proxy needs, or is this intended to be permanent? We could use this, but I'm concerned about introducing golang into our stack. I've played with it but haven't done serious production work with it. The upstream project hasn't gotten much use (16 bugs open / 28 closed), so we might well hit undocumented bugs. What's your comfort level with golang? Are there people on the wunderbar team who would be willing / have time to fix bugs? How does this fit into our broader strategy?

**Comment by chuckharmston** [comment]
> Is this a temporary addition until we solve the shared image proxy needs, or is this intended to be permanent?

Definitely temporary; this isn't a need that is going to go away as we continue experimenting with content in Firefox, and we'll eventually need a shared solution. I flagged that to @clouserw early on, and was told to use something off-the-shelf or as-a-service.

> We could use this, but I'm concerned about introducing golang into our stack. 

I'm not particularly comfortable adding Golang to our stack either, but this is a very cheap way for us to be kinder to our users in the interim. I'd be happy to look into other options, but whatever we do will need to be containerized, and the requirements + questions for ops won't change.

**Comment by chuckharmston** [comment]
Worth noting that the CDN would be nice here; it'd let us use some solutions in more familiar languages (e.g. [pilbox](https://github.com/agschwender/pilbox)) that don't cache. Unless the upstream cache (I'd imagine that we're using [`http_proxy`](http://nginx.org/en/docs/http/ngx_http_proxy_module.html)? Confirmation on that would be appreciated) would be able to handle that, which might be good enough.

**Comment by relud** [comment]
I think what would be easiest on my end would be to wait until @mostlygeek gets back from PTO and discuss then.

**Comment by chuckharmston** [comment]
Totally fine by me. Thanks!

File: conf/imageproxy/run.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+/go/bin/imageproxy \
```
Comment (mostlygeek) [inline conf/imageproxy/run.sh]: should be `exec /go/bin/imageproxy`

**Comment by mostlygeek** [comment]
Why this over the something like imgix mentioned in #104?

Also we should probably skip the S3 bucket cache. Instead, have nginx cache images in its cache which would be much times faster (latency wise) than S3 and also wouldn't require an additional backing service. 

Also the S3 code doesn't support STS which is how we pass credentials to access AWS resources to servers. [This bit here](https://github.com/willnorris/imageproxy/blob/master/vendor/sourcegraph.com/sourcegraph/s3cache/s3cache.go#L120-L121).

**Comment by chuckharmston** [comment]
> Why this over the something like imgix mentioned in #104?

Because we're trying to avoid passing user IP addresses to third parties. It additionally doesn't take advantage of any pipelining or multiplexing we end up implementing, which is pretty important for performance if we're sending over dozens of 16x16 images.

> Also we should probably skip the S3 bucket cache.

If you can stick this behind nginx, that works for me. 

Does this method of generating a container work for you? How would you like it integrated with the build process?

**Comment by chuckharmston** [comment]
@mostlygeek ping on the above ^

**Comment by mostlygeek** [comment]
I think this should be the architecture:

```
(net) --> (ELB) --> (nginx) --(http)--> (universal search) --> (image resizer)
```

What universal search should do is 

* request the image from the image resizer and have success/error logic
* pipe the image back to nginx with additional `Cache-Control` headers 
* nginx will cache things

The H2/pipelining/multiplexing stuff is decoupled from how images are requested, resized and cached. 

What will happen if we deploy H2 is: 

```
(ELB) --(tcp)--> (h2|nginx) --(http/1.1)--> (universal search)
```
Universal Search always gets HTTP/1.1 requests. Nothing will change on the right side of nginx.

**Comment by chuckharmston** [comment]
@mostlygeek I'm not exactly sure what you mean here. Are you suggesting that we rewrite the image proxy as part of the application, abandoning this pull request? Or that we only make the attached container accessible via private DNS, and let all the negotiation happen in the application?

In the former case, I don't think we have the engineering resources to do that. In the latter case, I don't see the benefits of tightly coupling the image proxy to our application.

> The H2/pipelining/multiplexing stuff is decoupled from how images are requested, resized and cached. 

In either of the cases, how is different than if we stand this image up behind a reverse proxy, e.g.

```
(net) --> (ELB) --> (nginx) --(http)--> (universal search)
                          \ --(http)--> (image resizer)
```

**Comment by mostlygeek** [comment]
I am suggestion that the app should abstract the logic on resizing/image generation. 

Implementing the proxy as you suggest would give anybody the ability to craft requests that reach out to the Internet. This opens up lots of liability and security issues. 

* Can the service be used to attack other services? 
* Can somebody DOS our servers cause they spending all their time resizing images? 
* etc. 

It is better to never open that door. 

How much effort  do you estimate to add an `imageproxy/<url>` endpoint that:

* normalizes / validates the URL 
* download the image from the resizer service 
* pipe the image back with `Cache-Control` headers? 

Embed.ly has an image resizing service (http://docs.embed.ly/docs/resize) that seems to do what we're looking for here.

**Pull request closed without merging**